presto-main/src/main/java/com/facebook/presto/type/DecimalCasts.java

```diff
@@ -419,10 +446,12 @@ public static long shortDecimalToFloat(long decimal, long precision, long scale,
     @UsedByGeneratedCode
     public static long longDecimalToFloat(Slice decimal, long precision, long scale, long tenToScale)
     {
-        // TODO: optimize
-        BigInteger decimalBigInteger = decodeUnscaledValue(decimal);
-        BigDecimal bigDecimal = new BigDecimal(decimalBigInteger, (int) scale);
-        return floatToRawIntBits(bigDecimal.floatValue());
+        // If both decimal and scale can be represented exactly in float then compute rescaled and rounded result directly in float.
+        if (scale < FLOAT_10_POW.length && compareUnsigned(decimal, MAX_EXACT_FLOAT) <= 0) {
+            return floatToRawIntBits((float) unscaledDecimalToUnscaledLongUnsafe(decimal) / FLOAT_10_POW[(int) scale]);
+        }
+
+        return floatToRawIntBits(parseFloat(Decimals.toString(decimal, (int) scale)));
```

Comment:
add comment:

`TODO: optimize and convert directly to double in similar fashion as in double to decimal casts`